netty/src/main/java/io/grpc/netty/WriteQueue.java

```diff
@@ -115,26 +105,31 @@ void enqueue(Runnable runnable, boolean flush) {
    * Process the queue of commands and dispatch them to the stream. This method is only
    * called in the event loop
    */
-  private void flush() {
+  @Override
+  @SuppressWarnings("ShortCircuitBoolean")
+  public void run() {
+    int writes = 0;
+    QueuedCommand cmd;
+    boolean wasWritable = channel.isWritable();
     try {
-      QueuedCommand cmd;
-      int i = 0;
-      boolean flushedOnce = false;
-      while ((cmd = queue.poll()) != null) {
-        cmd.run(channel);
-        if (++i == DEQUE_CHUNK_SIZE) {
-          i = 0;
-          // Flush each chunk so we are releasing buffers periodically. In theory this loop
-          // might never end as new events are continuously added to the queue, if we never
-          // flushed in that case we would be guaranteed to OOM.
-          channel.flush();
-          flushedOnce = true;
+      do {
+        while ((cmd = queue.poll()) != null) {
+          cmd.run(channel);
+          // There are two reasons to leave the loop:
+          // 1.  The channel's writability changed.
+          // 2.  We've done a lot of writes.
+          // Either way, we want to try another flush.  Use the non shortcircuiting | operator to
+          // make sure
+          if ((wasWritable != (wasWritable = channel.isWritable())) | (++writes == FLUSH_LIMIT)) {
+            // Always reset buffered writes, even in the good case.  We may have just exited the
```

Comment:
good case and "bad flush" aren't really defined. How about just do the `writes = 0` immediately before/after the `flush()`?